m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+                (CachedMetric) allocateCounter(name, durationTags)
+            ));
+        }
+
+        return new CachedHistogramImpl(
+            name,
+            tags,
+            buckets,
+            cachedValueBuckets,
+            cachedDurationBuckets
+        );
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            // A null metric signifies to processors that a flush was called,
+            // which causes the metric queue to be cleared
+            metricQueue.put(new SizedMetric(null, 0));
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private List<Metric> flush(List<Metric> metrics) throws TException {
+        TException exception = null;
+
+        synchronized (metricBatch) {
+            metricBatch.setMetrics(metrics);
+
+            try {
+                client.emitMetricBatch(metricBatch);
+            } catch (TException e) {
+                // Store exception to throw later after clean up.
+                // Don't want to put everything in a finally block here,
+                // which will hold the lock unnecessarily long
+                exception = e;
+            }
+
+            metricBatch.setMetrics(null);
+        }
+
+        resourcePool.releaseShallowMetrics(metrics);
+
+        metrics.clear();
+
+        if (exception != null) {
+            throw exception;
+        }
+
+        return metrics;
+    }
+
+    @Override
+    public void close() {
+        // Wait (block) until all processors to return from flushing
+        phaser.arriveAndAwaitAdvance();
+    }
+
+    private Metric newMetric(String name, Map<String, String> tags, MetricType metricType) {
+        Metric metric = resourcePool.getMetric();
+        metric.setName(name);
+
+        metric.setTags(addMetricTagsToSet(tags));
+
+        metric.setTimestamp(Long.MAX_VALUE);
+
+        MetricValue metricValue = resourcePool.getMetricValue();
+
+        switch (metricType) {
+            case COUNTER:
+                CountValue countValue = resourcePool.getCountValue();
+                countValue.setI64Value(Long.MAX_VALUE);
+                metricValue.setCount(countValue);
+                break;
+            case GAUGE:
+                GaugeValue gaugeValue = resourcePool.getGaugeValue();
+                gaugeValue.setDValue(Double.MAX_VALUE);
+                metricValue.setGauge(gaugeValue);
+                break;
+            case TIMER:
+                TimerValue timerValue = resourcePool.getTimerValue();
+                timerValue.setI64Value(Long.MAX_VALUE);
+                metricValue.setTimer(timerValue);
+                break;
+        }
+
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private Set<MetricTag> addMetricTagsToSet(Map<String, String> tags) {
```

Comment:
nit: rename to indicate you're creating a set instead of adding to a provided set (e.g. `toMetricTagSet`)